components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/MobileNumberVerificationHandler.java

```diff
@@ -0,0 +1,380 @@
+     * @param user              User.
+     * @throws IdentityEventException
+     */
+    private void preSetUserClaimOnMobileNumberUpdate(Map<String, String> claims, UserStoreManager userStoreManager,
+                                                     User user) throws IdentityEventException {
+
+        if (IdentityRecoveryConstants.SkipMobileNumberVerificationOnUpdateStates.SKIP_ON_CONFIRM.toString().equals
+                (Utils.getThreadLocalToSkipSendingSmsOtpVerificationOnUpdate())) {
+            // Not required to handle in this handler.
+            return;
+        }
+
+        if (Utils.getThreadLocalToSkipSendingSmsOtpVerificationOnUpdate() != null) {
+            Utils.unsetThreadLocalToSkipSendingSmsOtpVerificationOnUpdate();
+        }
+
+        if (MapUtils.isEmpty(claims)) {
+            // Not required to handle in this handler.
+            Utils.setThreadLocalToSkipSendingSmsOtpVerificationOnUpdate(IdentityRecoveryConstants
+                    .SkipMobileNumberVerificationOnUpdateStates.SKIP_ON_INAPPLICABLE_CLAIMS.toString());
+            return;
+        }
+
+        String mobileNumber = claims.get(IdentityRecoveryConstants.MOBILE_NUMBER_CLAIM);
+
+        if (StringUtils.isNotBlank(mobileNumber)) {
+            String existingMobileNumber;
+            String username = user.getUserName();
+            try {
+                existingMobileNumber = userStoreManager.getUserClaimValue(username, IdentityRecoveryConstants.
+                        MOBILE_NUMBER_CLAIM, null);
+            } catch (UserStoreException e) {
+                String error = String.format("Error occurred while retrieving existing mobile number for user: %s in " +
+                        "domain : %s and user store : %s", username, user.getTenantDomain(), user.getUserStoreDomain());
+                throw new IdentityEventException(error, e);
+            }
+
+            if (StringUtils.equals(mobileNumber, existingMobileNumber)) {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("The mobile number to be updated: %s is same as the existing mobile " +
+                            "number for user: %s in domain %s. Hence an SMS OTP verification will not be " +
+                            "triggered.", mobileNumber, username, user.getTenantDomain()));
+                }
+                Utils.setThreadLocalToSkipSendingSmsOtpVerificationOnUpdate(IdentityRecoveryConstants
+                        .SkipMobileNumberVerificationOnUpdateStates.SKIP_ON_EXISTING_MOBILE_NUM.toString());
+                invalidatePendingMobileVerification(user, userStoreManager, claims);
+                return;
+            }
+            claims.put(IdentityRecoveryConstants.MOBILE_NUMBER_PENDING_VALUE_CLAIM, mobileNumber);
+            claims.remove(IdentityRecoveryConstants.MOBILE_NUMBER_CLAIM);
+        } else {
+            claims.put(IdentityRecoveryConstants.MOBILE_NUMBER_PENDING_VALUE_CLAIM, StringUtils.EMPTY);
```

Comment:
Why do we need this? Won't this remove the MOBILE_NUMBER_PENDING_VALUE_CLAIM value if the request is an update request which doesn't contain the mobile number claim and hence break the expected behaviour?

Reply:
Addressed in https://github.com/wso2-extensions/identity-governance/pull/414/commits/b5f0c45e2ce7cb40a24f7a051acc52455755cce4